File: src/main/scala/com/abaddon83/cardsGames/burracoGames/commands/DropScaleCmd.scala
package com.abaddon83.cardsGames.burracoGames.commands

import java.util.UUID

import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.initialised.playerInGames.BurracoScale
import com.abaddon83.cardsGames.burracoGames.ports.GameRepositoryPort
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.cardsGames.shares.players.PlayerIdentity
import com.abaddon83.libs.cqs.commands.{Command, CommandHandler}

import scala.concurrent.{Await, Future}

case class DropScaleCmd(
                     gameIdentity: GameIdentity,
                     playerIdentity: PlayerIdentity,
                     scale: BurracoScale
                   ) extends Command {
  override protected val requestId: UUID = UUID.randomUUID()
}

class DropScaleHandler(
                       gameRepositoryPort: GameRepositoryPort
                     )
                     (implicit val ec: scala.concurrent.ExecutionContext)
  extends CommandHandler[DropScaleCmd] {

  override def handleAsync(command: DropScaleCmd): Future[Unit] = {
    for {
      gameExecution <- gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity)
      gameExecutionUpdated = gameExecution.dropOnTableAScale(playerIdentity = command.playerIdentity, scale = command.scale)
    } yield gameRepositoryPort.save(gameExecutionUpdated)
  }

  override def handle(command: DropScaleCmd): Unit = {
    import scala.concurrent.duration._
    import scala.language.postfixOps
    val gameExecution = Await.result(gameRepositoryPort.findBurracoGameInitialisedTurnExecutionBy(gameIdentity = command.gameIdentity), 500 millis)
    val gameExecutionUpdated = gameExecution.dropOnTableAScale(playerIdentity = command.playerIdentity, scale = command.scale)
    gameRepositoryPort.save(gameExecutionUpdated)
  }

}

File: src/main/kotlin/com/abaddon83/burracoGame/commands/DropScaleCmd.kt
package com.abaddon83.burracoGame.commands

import com.abaddon83.burracoGame.domainModels.BurracoGame
import com.abaddon83.burracoGame.domainModels.BurracoScale
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.decks.Card
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.burracoGame.shared.players.PlayerIdentity
import com.abaddon83.utils.cqs.commands.Command
import com.abaddon83.utils.cqs.commands.CommandHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.sql.Timestamp
import java.util.*

data class DropScaleCmd(
        val gameIdentity: GameIdentity,
        val playerIdentity: PlayerIdentity,
        val scale: BurracoScale,
        override val commandId: UUID,
        override val commandTimeStamp: Timestamp): Command {

    constructor(gameIdentity: GameIdentity,playerIdentity: PlayerIdentity,scale: BurracoScale): this(
            gameIdentity = gameIdentity,
            playerIdentity = playerIdentity,
            scale = scale,
            commandId = UUID.randomUUID(),
            commandTimeStamp = Timestamp(System.currentTimeMillis()))
}

class DropScaleHandler() : CommandHandler<DropScaleCmd>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(command: DropScaleCmd) {
        val gameExecution = runBlocking {repository.findBurracoGameExecutionTurnExecutionBy(gameIdentity = command.gameIdentity)}
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameExecution = gameExecution.dropOnTableAScale(playerIdentity = command.playerIdentity, scale = command.scale)
        repository.save(updatedGameExecution)
    }

    override suspend fun handleAsync(command: DropScaleCmd) {
        val gameExecution = repository.findBurracoGameExecutionTurnExecutionBy(gameIdentity = command.gameIdentity)
        check(gameExecution != null) {"GameIdentity ${command.gameIdentity} not found exist"}
        val updatedGameExecution = gameExecution.dropOnTableAScale(playerIdentity = command.playerIdentity, scale = command.scale)
        repository.save(updatedGameExecution)
    }
}